### ISSUE TITLE
Not really working... at all.

### ISSUE THREAD

**Issue by flber** [usysrc/iga#1]
As far as I can tell, the algorithm currently pseudo-randomly places the input tiles into the given space. I'm not sure how to fix it, or even what's broken.

**Comment by Skaruts** [comment]
@flber 
It's just that Love2d changed how it exposes image data in the meanwhile. `img:getData()` has been deprecated. 
It's easy to fix. You just need to change this:
https://github.com/usysrc/iga/blob/964ab9f4123935949e2e393e047c7bec4eee18a4/main.lua#L7-L11
To this:
```lua
img_data = love.image.newImageData("input/input_10.png")
local ow, oh = 100,100
iga:geninit(img_data, ow,ow, img_data:getWidth(), img_data:getHeight(), 3)
```

And then in [iga:geninit()](https://github.com/usysrc/iga/blob/master/iga.lua#L42) make these three changes:

```lua
function iga:geninit(img_data, ow,oh,iw,ih,N_)  -- line 42
	[...]

	-- remove this line (it's not needed anymore)
	local data = img:getData()	-- line 83

	for i=0,input_width-1 do
		for j=0,input_height-1 do
			[...]
			local r,g,b = img_data:getPixel(i,j)   -- line 88
```

If I didn't forget anything, it should work now. 
Although I should point out that for some odd reason this program has no way to quit or close. You can remove `{borderless=true}` from **main.lua**, and/or add a key to quit:
```lua
function love.keypressed(k)
	if k == "escape" then love.event.quit() end
end
```

**Comment by flber** [comment]
Well that's sort of odd, I can't see the email I just got through github...

Anyway, after patching iga to [this state](https://mineralexistence.com/iga.diff), I still get a scrambled output.
I think I fixed everything mentioned, but I could be wrong. And just as a sanity check, here's the message I got:
```
@flber
It's just that Love2d changed how it exposes image data in the meanwhile. img:getData() has been deprecated.
It's easy to fix. You just need to change this:
https://github.com/usysrc/iga/blob/964ab9f4123935949e2e393e047c7bec4eee18a4/main.lua#L7-L11
To this:

img_data = love.image.newImageData("input/input_10.png")
local ow, oh = 100,100
iga:geninit(img_data, ow,ow, img_data:getWidth(), img_data:getHeight(), 3)
And then in iga:geninit() make these three changes:

function iga:geninit(img_data, ow,oh,iw,ih,N_)  -- line 42
	[...]

	-- remove this line (it's not needed anymore)
	local data = img:getData()	-- line 83

	for i=0,input_width-1 do
		for j=0,input_height-1 do
			[...]
			local r,g,b = img_data:getPixel(i,j)   -- line 88
If I didn't forget anything, it should work now.
Although I should point out that for some odd reason this program has no way to quit or close. You can remove {borderless=true} from main.lua, and/or add a key to quit:

function love.keypressed(k)
	if k == "escape" then love.event.quit() end
end
```

**Comment by Skaruts** [comment]
@flber sorry, I deleted my reply because I thought I had totally misunderstood you. At first I thought you'd be getting the error about `img:getPixel()`, but then I noticed the algorithm in this port isn't working properly either. It's not producing the same results as the original. Re-reading your post after that made me think that that's what you meant, and not what I had interpreted before. I was under too much sleep deprivation... :(

So yea, this is indeed not working properly, anyway. I tested it with the sample images from the original C# repo, and the results are also screwed up. 

The original lua port seems to work fine, though, at least when running its own executable. I haven't tried porting it to love2d, but it might be worth looking into.

Again, sorry about the confusion.

**Comment by flber** [comment]
@Skaruts ah well, no problem

**Comment by Skaruts** [comment]
@flber well I found the issue causing the scrambled output: colors have also been changed in love2d in the meanwhile. They use floats 0-1 now, and this was made when it used 0-255. 
So to make this work better you only need to convert them:

```lua
-- iga.lua -- line 89: multiply components by 255
local colour = string.format("%03d%03d%03d",r*255,g*255,b*255)
```
```lua
-- main.lua -- line 26: convert back to floats, by diving them
love.graphics.setColor(tonumber(r)/255,tonumber(g)/255,tonumber(b)/255)
```
However, the results are still not looking like the original lua port. There are still black spots and plenty of artifacts. For example, the very first sample (a white cross) in the original port generates a straight grid, while this one generates a very disjointed resemblance of a grid.
